tools/ci/pr_preview.py

```diff
@@ -0,0 +1,396 @@
+#!/usr/bin/env python
+
+# The service provided by this script is not critical, but it shares a GitHub
+# API request quota with critical services. For this reason, all requests to
+# the GitHub API are preceded by a "guard" which verifies that the subsequent
+# request will not deplete the shared quota.
+#
+# In effect, this script will fail rather than interfere with the operation of
+# critical services.
+
+import argparse
+import json
+import logging
+import os
+import subprocess
+import time
+
+import requests
+
+# The ratio of "requests remaining" to "total request quota" below which this
+# script should refuse to interact with the GitHub.com API
+API_RATE_LIMIT_THRESHOLD = 0.2
+# The GitHub Pull Request label which indicates that a Pull Request is expected
+# to be actively mirrored by the preview server
+LABEL = 'pull-request-has-preview'
```

Comment:
I'd like to not use a label here. When it was used, it cluttered up any PR listing quite badly, and when we're using deployments instead it ought not be needed.

Reply:
The GitHub Issue Label isn't a vestige from the prior implementation; it's the user-facing mechanism for enabling previews from untrusted forks. That's important even in a solution that uses GitHub Deployments. To me, it's an improvement over the comment-based approach offered by the w3c-test.org implementation.

We could use the label *only* for that purpose, so that only humans apply it. One drawback is project collaborators won't be able to opt out of the automatic preview (under the solution proposed here, they would do this by removing the automatically-assigned Label). Opting out is a feature of the w3c-test.org implementation, but I doubt it's a highly desired one, so maybe that's acceptable.

If we do limit the usage in this way, I'd like to change the name to "safelisted-for-preview".

Currently, the wpt.fyi system publishes previews for all refs that exist in two namespaces: `refs/prs-open/*` and `refs/prs-labeled-for-preview/*`. With this change, we'll certainly want to rework that. The simplest would be to rename the latter to `refs/prs-trusted-for-preview/*`, though that promotes the concept of "trust" in a way that couples the two projects. It would be more hygienic to work from *three* namespaces: `refs/prs-open/*`, `refs/prs-safelisted/*`, and `refs/prs-from-collaborators/*`.

So while removing automated application of the Label is fairly straightforward, it has ramifications that require a few hours of work. I'll hold off on that until there's agreement on the topic.

Reply:
A "safelisted-for-preview" label that's only used to trigger mirroring where it otherwise wouldn't happen sounds good to me.

Since the refs are just to communicate to wptpr.live what it should do and it doesn't need to know why a PR should be deployed, won't two sets of refs still suffice? Just rename `refs/prs-labeled-for-preview/*` to `refs/prs-deployable/*` or something?

Reply:
> Since the refs are just to communicate to wptpr.live what it should do and it doesn't need to know why a PR should be deployed, won't two sets of refs still suffice?

That's true, and going further, we only technically need one ref. Something like `refs/prs-open-and-trusted/*` would be sufficient for wptpr.live to do its job.

Great minds think alike! That's how I originally designed [the prior iteration of this task](https://github.com/web-platform-tests/wpt/pull/17680). @jgraham [pushed back on that, writing](https://github.com/web-platform-tests/wpt/pull/17680#issuecomment-510000469)

> it highly couples things to the exact requirements of live preview; we only get tags on items that are both open PRs and also have the right label set. We could decouple those things and create refs for open PRs and refs for things labelled for preview.

I agreed, which is why this iteration started with two ref namespaces. My suggestion to introduce a third namespace is an application of the same rationale.

Then again, while a distinct `prs-open` ref namespace might be useful elsewhere, I'm less confident that a `prs-from-collaborators` would be (and even less so about `prs-safelisted`). I don't want to increase complexity to reduce coupling if there's no practical benefit. @jgraham do you have an opinion?

Reply:
Not a strong opinion, but I think two is fine. "PRs that are open" seems like a clearly reusable thing, but trying to split the difference between "PRs that are trusted because they're from collaborators" and "PRs that are trusted because they were marked as such" seems like it might be overkill in terms of enabling other consumers.

Reply:
Thanks, @jgraham. I've updated this accordingly. To summarize: the GitHub Issue label in use is "safelisted-for-preview", and it is only referenced by this script (that is, never introduced) in order to determine when a submission from a non-Collaborator should be mirrored. The underlying ref namespace is `refs/prs-trusted-for-preview/*`.

...I think. This is arduous to verify because the patch still lacks automated tests. Now that three members of the "wpt core team" have participated in this discussion, and no one has objected, I think it's prudent to write some tests. I'll plan to get started on that tomorrow.